### PR TITLE
IGNITE-15234 Implement ByteBuf-based message packer and unpacker

### PR DESCRIPTION
**Pull request by ptupitsyn**
* Refactor `ClientMessagePacker` and `ClientMessageUnpacker` to work with Netty `BytBuf` directly, without any wrappers.
* Remove inheritance from library implementation.
* Remove unnecessary `ByteBuffer` allocations.

**Benchmarks Before:**

```
Benchmark                                 Mode  Cnt        Score        Error  Units
ClientMessagePackerBenchmark.packInt     thrpt    3  2868266.504 ± 173548.310  ops/s
ClientMessagePackerBenchmark.packString  thrpt    3  2663375.726 ± 190027.949  ops/s
ClientMessagePackerBenchmark.packUuid    thrpt    3  2816744.141 ± 281943.169  ops/s
```

**Benchmarks After:**

```
Benchmark                                 Mode  Cnt          Score          Error  Units
ClientMessagePackerBenchmark.packInt     thrpt    3  389568322.545 ± 48247870.394  ops/s   (135x)
ClientMessagePackerBenchmark.packString  thrpt    3   30113891.585 ±  1781762.180  ops/s   (11x)
ClientMessagePackerBenchmark.packUuid    thrpt    3  331495906.854 ± 12134046.484  ops/s   (117x)
```

NOTE: Those benchmarks include `ClientMessagePacker` constructor call.

### CODE REVIEW

**Review by AMashenkov — APPROVED** [review]

File: modules/client-common/src/main/java/org/apache/ignite/internal/client/proto/ClientMessagePacker.java [hunk 1]
```diff
@@ -36,637 +34,748 @@
...
-
+        
         packArrayHeader(args.length);
-
+        
         for (Object arg : args) {
             if (arg == null) {
                 packNil();
-
+
```
Comment (isapego) [inline modules/client-common/src/main/java/org/apache/ignite/internal/client/proto/ClientMessagePacker.java:773]: Maybe you should tweak your IDE a bit to remove trailing whitespaces :)

File: modules/client-handler/src/integrationTest/java/org/apache/ignite/client/handler/ItClientHandlerTest.java [hunk 2]
```diff
@@ -167,7 +167,7 @@ void testHandshakeInvalidVersionReturnsError() throws Exception {
             final var err = unpacker.unpackString();
             
             assertArrayEquals(MAGIC, magic);
-            assertEquals(31, len);
+            assertEquals(32, len);
```
Comment (isapego) [inline modules/client-handler/src/integrationTest/java/org/apache/ignite/client/handler/ItClientHandlerTest.java:170]: Why this change occurred?

**Review by isapego — APPROVED** [review]

File: modules/client-handler/src/integrationTest/java/org/apache/ignite/client/handler/ItClientHandlerTest.java (hunk 2 above)
Comment (ptupitsyn) [reply modules/client-handler/src/integrationTest/java/org/apache/ignite/client/handler/ItClientHandlerTest.java:170]: Please see comments in `ClientMessagePacker#packString`. Basically we reserve string header size pessimistically to avoid extra string scan. It is better to write 1 extra byte than to spend time on precise utf8 byte length calculation.

The code from the library that we've used before was doing `byte[] bytes = str.getBytes()` and then copied those bytes to the target buffer, which involves an extra allocation and an extra copy.

File: modules/client-common/src/main/java/org/apache/ignite/internal/client/proto/ClientMessagePacker.java (hunk 1 above)
Comment (ptupitsyn) [reply modules/client-common/src/main/java/org/apache/ignite/internal/client/proto/ClientMessagePacker.java]: I have this enabled, but I've used auto-reformat to conform with the new code style and that's where those spaces come from.

Fixed.

File: modules/client-handler/src/integrationTest/java/org/apache/ignite/client/handler/ItClientHandlerTest.java (hunk 2 above)
Comment (isapego) [reply modules/client-handler/src/integrationTest/java/org/apache/ignite/client/handler/ItClientHandlerTest.java:170]: Okay then, I just wanted to make sure it won't break compatibility with other implementation

Comment (ptupitsyn) [reply modules/client-handler/src/integrationTest/java/org/apache/ignite/client/handler/ItClientHandlerTest.java:170]: Yes, this is compatible. We can use bigger varint format even if the number if small (e.g. number 5 can be encoded as 1 byte, but it is allowed to use 2-byte or 4-byte formats too).
